### PR TITLE
Henry/conditional formatting

### LINKED ISSUE #18: Conditional formatting.
Would it be possible to support conditional formatting in this widget, depending on the number of days left? Thanks.

### PR DESCRIPTION
**Pull request by henrybeen**
Thank you for your pull request!
By completing this pull request, you agree to the [Contributing License Agreement](https://github.com/ALM-Rangers/Countdown-Widget-Extension/blob/master/.github/CLA.md).

Changes proposed in this pull request:  
- Some reorganizing in countdownWidget.ts to improve the visibility of control flow
- Fixes #18 - Conditional formatting

@ALM-Rangers/countdownwidget

afbeelding

### CODE REVIEW

**Comment by jessehouwing** [comment]
Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---

*[CountdownWidget/CountdownWidget/src/configuration.ts, line 146 at r1](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32#-LA8HDOnD55_XYnKRI_C:-LA8HDOnD55_XYnKRI_D:bp34fy1) ([raw file](https://github.com/alm-rangers/countdown-widget-extension/blob/82594c7ac4463dc09e0204baca4678c7d30fe816/CountdownWidget/CountdownWidget/src/configuration.ts#L146)):*
> ```XML
> 		colorSettings.color = (settings && settings.backgroundColorHoursColor)
> 			? settings.backgroundColorHoursColor
> 			: "red";
> ```

Why not make this a `settings.defaultBackgroundColor?`

---

*[CountdownWidget/CountdownWidget/src/countdownWidget.ts, line 175 at r1](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32#-LA8HgecDhkUwNgo0Oex:-LA8HgecDhkUwNgo0Oey:bfaih8w) ([raw file](https://github.com/alm-rangers/countdown-widget-extension/blob/82594c7ac4463dc09e0204baca4678c7d30fe816/CountdownWidget/CountdownWidget/src/countdownWidget.ts#L175)):*
> ```XML
> 		$title.text(this.currentSettings.name);
> 
> 		$countDownBody.css("background-color", this.renderValues.backgroundColor);
> ```

This requires 'unsafe-inline' for Content Security Policy. It's probably better to manipulate the DOM element directly instead of overriding the css value.

---

*[CountdownWidget/CountdownWidget/static/css/app.css, line 44 at r1](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32#-LA8Gfmm2i7onH5LUnWo:-LA8Gfmm2i7onH5LUnWp:bkxfwql) ([raw file](https://github.com/alm-rangers/countdown-widget-extension/blob/82594c7ac4463dc09e0204baca4678c7d30fe816/CountdownWidget/CountdownWidget/static/css/app.css#L44)):*
>> foreground-picker

---

*[CountdownWidget/CountdownWidget/static/css/app.css, line 70 at r1](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32#-LA8I0Sc4VS1uTPzwtXW:-LA8I0Sc4VS1uTPzwtXX:b821j7c) ([raw file](https://github.com/alm-rangers/countdown-widget-extension/blob/82594c7ac4463dc09e0204baca4678c7d30fe816/CountdownWidget/CountdownWidget/static/css/app.css#L70)):*
> ```CSS
> }
> 
> .indented {
> ```

Personally, don't like CSS classes that have the name of what they do instead of the name of what they're intended for. 

What is getting indented here. Why is that so, is there a better name?

---

*[CountdownWidget/CountdownWidget/static/css/app.css, line 76 at r1](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32#-LA8ID_89X-pxWfrFLHb:-LA8ID_89X-pxWfrFLHc:bcpoun9) ([raw file](https://github.com/alm-rangers/countdown-widget-extension/blob/82594c7ac4463dc09e0204baca4678c7d30fe816/CountdownWidget/CountdownWidget/static/css/app.css#L76)):*
> ```CSS
> }
> 
> .inputwrapper {
> ```

input-wrapper? or inputWrapper?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32)*

**Comment by jessehouwing** [comment]
FYI few remarks. Plus, seems that master is again ahead of this branch. So a new rebase/merge is needed.

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32#-:-LA8IjJg1RjZ6r_kMbZ0:b-3djtsj)*

**Comment by jessehouwing** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[CountdownWidget/CountdownWidget/package.json, line 42 at r1](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32#-LA8JxmB6dzPocPhHZb7:-LA8JxmB6dzPocPhHZb8:b6l2vdn) ([raw file](https://github.com/alm-rangers/countdown-widget-extension/blob/82594c7ac4463dc09e0204baca4678c7d30fe816/CountdownWidget/CountdownWidget/package.json#L42)):*
> ```JSON
>   },
>   "dependencies": {
>     "@types/jasmine": "^2.5.53",
> ```

Seems to be a change that was not really intentional... Just moved the lines.

---

*[CountdownWidget/CountdownWidget/package.json, line 44 at r1](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32#-LA8Jx-0BThkxisleCEV:-LA8Jx-0BThkxisleCEW:bb7407z) ([raw file](https://github.com/alm-rangers/countdown-widget-extension/blob/612c3e0fbd1c5985d70edd88d1bb0150bf55fc03/CountdownWidget/CountdownWidget/package.json#L44)):*
> ```JSON
>     "@types/jquery": "^2.0.34",
>     "@types/q": "0.0.32",
>     "@types/jasmine": "^2.5.53",
> ```

Seem

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/alm-rangers/countdown-widget-extension/32)*

File: CountdownWidget/CountdownWidget/src/countdownWidget.ts [hunk 1]
```diff
@@ -37,174 +43,174 @@ export class CountdownWiget {
...
+			if (this.currentSettings.skipNonWorkingDays) {
+				this.workingDays = teamSettings.workingDays;
+			}
+		});
+	}
+
+	private determineRenderValues() {
+		if (this.isSprintEndWidget) {
+			return this.determineSprintWidgetRenderValues();
+		} else {
```
Comment (d3r3kk) [inline CountdownWidget/CountdownWidget/src/countdownWidget.ts]: Probably don't need the else clause?

File: CountdownWidget/CountdownWidget/src/CountdownResult.ts [hunk 2]
```diff
@@ -36,6 +36,18 @@ export class CountdownResult {
...
+		if (this.unit === unit) {
+			return this.value < threshold;
+		}
+
+		return false;
+	}
 }
 
-export enum Unit { Years, Days, Hours, Minutes, Seconds, Invalid }
+export enum Unit { Invalid = 0, Seconds = 1, Minutes = 2, Hours = 3, Days = 4, Years = 5}
```
Comment (d3r3kk) [inline CountdownWidget/CountdownWidget/src/CountdownResult.ts]: Very nice, much clearer to make this explicit I think. 
However, does it make any difference in how it has been used up to this point? Is there any places where > < is being used to compare between them?

File: CountdownWidget/CountdownWidget/src/countdownWidget.ts [hunk 3]
```diff
@@ -37,174 +43,174 @@ export class CountdownWiget {
...
+				})
+				.then(() => {
+					this.calculateCountdown();
+					this.calculateBackgroundColor();
+					this.renderWidget();
+				});
+
+			return this.WidgetHelpers.WidgetStatusHelper.Success();
+		} catch (e) {
+			// Satisfy the linter
```
Comment (d3r3kk) [inline CountdownWidget/CountdownWidget/src/countdownWidget.ts]: Any way to capture a log line here? Something we could use to troubleshoot if necessary? It seems a bit counter-intuitive to just swallow a potential problem and not capture something about it.

File: CountdownWidget/CountdownWidget/src/CountdownResult.ts (hunk 2 above)
Comment (kenmuse) [reply CountdownWidget/CountdownWidget/src/CountdownResult.ts]: Definitely like how easy it is to read the code. Are there any places where this.value can be large enough to be a different unit (for example, 90 minutes)?

**Pull request merged**